src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComActivator.cs

```diff
@@ -0,0 +1,301 @@
+                && cxt.InterfaceId != typeof(IClassFactory2).GUID)
+            {
+                throw new NotSupportedException();
+            }
+
+            string[] potentialAssemblies = cxt.ActivationAssemblyList ?? new string[0];
+            (Assembly classAssembly, Type classType) = FindClassAssemblyAndType(cxt.ClassId, potentialAssemblies);
+            return new BasicClassFactory(cxt.ClassId, classAssembly, classType);
+        }
+
+        /// <summary>
+        /// Internal entry point for unmanaged COM activation API from native code
+        /// </summary>
+        /// <param name="cxtInt">Reference to a <see cref="ComActivationContextInternal"/> instance</param>
+        public static int GetClassFactoryForTypeInternal(ref ComActivationContextInternal cxtInt)
+        {
+            if (IsLoggingEnabled())
+            {
+                Log(
+$@"{nameof(GetClassFactoryForTypeInternal)} arguments:
+    {cxtInt.ClassId}
+    {cxtInt.InterfaceId}
+    {cxtInt.AssemblyCount}
+    0x{cxtInt.AssemblyList.ToInt64():x}
+    0x{cxtInt.ClassFactoryDest.ToInt64():x}");
+            }
+
+            try
+            {
+                string[] potentialAssembies = CreateAssemblyArray(cxtInt.AssemblyCount, cxtInt.AssemblyList);
+
+                var cxt = new ComActivationContext()
+                {
+                    ClassId = cxtInt.ClassId,
+                    InterfaceId = cxtInt.InterfaceId,
+                    ActivationAssemblyList = potentialAssembies
+                };
+
+                object cf = GetClassFactoryForType(cxt);
+                IntPtr nativeIUnknown = Marshal.GetIUnknownForObject(cf);
+                Marshal.WriteIntPtr(cxtInt.ClassFactoryDest, nativeIUnknown);
+            }
+            catch (Exception e)
+            {
+                return e.HResult;
+            }
+
+            return 0;
+        }
+
+        private static bool IsLoggingEnabled()
+        {
+#if DEBUG
+            return true;
+#else
+            return false;
+#endif
+        }
+
+        private static void Log(string fmt, params object[] args)
+        {
+            // [TODO] Consider using FrameworkEventSource in release builds
+
+            Debug.WriteLine(fmt, args);
+         }
+
+        private static (Assembly assembly, Type type) FindClassAssemblyAndType(Guid clsid, string[] potentialAssembies)
+        {
+            // Determine what assembly the class is in
+            foreach (string assemPath in potentialAssembies)
+            {
+                Assembly assem;
+                string assemPathLocal = assemPath;
+
+                try
+                {
+                    string extMaybe = Path.GetExtension(assemPath);
+                    if (".manifest".Equals(extMaybe, StringComparison.OrdinalIgnoreCase))
+                    {
+                        assemPathLocal = Path.ChangeExtension(assemPath, ".dll");
+                    }
+
+                    assem = Assembly.LoadFrom(assemPathLocal);
```

Comment:
Loading a bunch of assemblies using LoadFrom is asking for problems. The dependencies of these assemblies may not be resolved correctly, and the app may not expect all of them to be loaded.

And then loading all types in these assemblies using GetTypes is performance trap.

Is this temporary hack?

Reply:
Yes. This is purely for satisfying test bring up. Once the API surface area is signed off, we need to do a lot more here. Figure out ALC and such along with probably trying to parse metadata instead of loading the assembly itself. Perhaps using `Assembly.ReflectionOnlyLoadFrom()` in the long run?

Reply:
I will be looking to @vitek-karas to keep my honest here :-) Is the `Assembly.ReflectionOnlyLoadFrom()` the right API to interrogate the types in an assembly without perturbing the running CLR instance?

Reply:
ReflectionOnlyLoadFrom is not supported in .NET Core.

Is there no way to avoid the linear search to find the right type? I think we should have the type name in some manifest to avoid the linear search.

Reply:
>Is there no way to avoid the linear search to find the right type? I think we should have the type name in some manifest to avoid the linear search.

There are options I still need to explore, but in this case that is really secondary. The real issue is finding which assembly in the activation context is the assembly with the type in question. In the example there is one assembly, but conceivably there could be N assemblies and each need to be queried to determine the one that provides the type.

Reply:
I guess the better question here is what is the appropriate way to read an assemblies metadata without loading the assembly into the current CLR?

Reply:
- Unmanaged: IMDInternalImport interface
- Managed: System.Reflection.Metadata https://github.com/dotnet/corefx/tree/master/src/System.Reflection.Metadata